### ISSUE TITLE
no-unused-vars option to allow unused vars in object destructuring with rest property

### ISSUE THREAD

**Issue by bryanrsmith** [eslint/eslint#4880]
I'd like an option for the `no-unused-vars` rule that would allow unused variables that are created as part of an object destructuring pattern that also includes an experimental rest property.

Use case: experimental object rest properties offer a convenient way to shallow clone an object while omitting certain properties. 
For example,

```js
let foo = { a: 1, b: 2, extra: { other: 'stuff' }};

// I want a shallow clone of `foo` that omits `extra`
let { extra, ...bar } = foo; // error-- `extra` is unused

return bar; // { a: 1, b: 2 }
```

This can be done now by naming the unused variables with a pattern (instead of using the shorthand) and configuring `no-unused-vars` with `varsIgnorePattern`. It's less convenient and it adds noise, though.

```js
let { extra: ignored, ...bar } = foo;
```

Rest/spread properties are still a stage 2 proposal, so it might be premature to add an option to a core rule. I figured I'd open for discussion anyways :-)

**Comment by eslintbot** [comment]
@bryanrsmith Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by nzakas** [comment]
Yeah, too early to add an exception. We don't make special affordances for experimental features, but feel free to follow up if it becomes official.

**Comment by bryanrsmith** [comment]
Thanks. Will do.

**Comment by oliviertassinari** [comment]
@bryanrsmith What's you work around regarding this point?
It seems that disabling the rule is the best option for Material-UI.

**Comment by bryanrsmith** [comment]
@oliviertassinari For now I used the `varsIgnorePattern` workaround described above.

**Comment by neoziro** [comment]
I have the same problem, is it possible to tweak using a plugin?

**Comment by oliviertassinari** [comment]
@neoziro I think that https://github.com/babel/eslint-plugin-babel would be the right location to add the support for this feature.

**Comment by bryanrsmith** [comment]
I've written a (hopefully temporary) plugin for this at https://github.com/bryanrsmith/eslint-plugin-no-unused-vars-rest. It basically just monkey patches the option into the existing core rule.

I'd be happy to submit a PR to eslint-plugin-babel if everyone would prefer that. It's a little odd, though, because it doesn't require babel-eslint. Espree's `experimentalObjectRestSpread` option covers it. What do you folks think?

**Comment by mnpenner** [comment]
Thank you @bryanrsmith. That's exactly what I was looking for! Was just about to disable `no-unused-vars` altogether because of this.

**Comment by uniqname** [comment]
Is this not already a valid, current spec compliant pattern when used with array destructuring?

`const [head, ...tail] = someArray;`

**Comment by mmcgahan** [comment]
@uniqname this is for [Object destructuring assignment](https://github.com/sebmarkbage/ecmascript-rest-spread), which is still a Stage 3 proposal, separate from Array destructuring, which is an ES2015 standard.

**Comment by uniqname** [comment]
@mmcgahan I understand that, but the `no-unused-vars` option does not appear to support this pattern for _array_ destructuring either. I can open a separate issue if that is preferable, unless I am mistaken and this option _is_ supported for this pattern with array destructuring.

**Comment by not-an-aardvark** [comment]
With array destructuring, you can use this as a workaround:

```js
const [ , ...tail] = someArray;
```

This isn't possible with object destructuring.

**Comment by uniqname** [comment]
Interesting. I'm not sure I would prefer that to a named destructured variable that is optionally ignored by islint, but at least the pattern is supported. Thanks @not-an-aardvark

**Comment by jameswomack** [comment]
For posterity's sake:
https://eslint.org/docs/rules/no-unused-vars#ignorerestsiblings

The `ignorerestsiblings` option for the `no-unused-var` rule accounts for this use case now.
